---
title: "R Notebook"
output: html_notebook
---

# Setup
```{r, message=FALSE}
library(ggplot2)
library(Biostrings)
library('seqinr')
library('spgs')  # reverseComplement("actg")
library("viridis")

library('egg')  #ggarrange
source("/Users/anna/OneDrive/pushkin/cryptic/nanopore/scripts/synteny_infer.R")

path.base = '/Volumes/Samsung_T5/vienn/tair/'
path.query = '/Volumes/Samsung_T5/vienn/renamed/'
path.blast.res <- '/Volumes/Samsung_T5/vienn/blast_res/'
path.t <- '/Volumes/Samsung_T5/vienn/alignments/'
path.figures <- '/Volumes/Samsung_T5/vienn/figures/'
path.snps <- '/Volumes/Samsung_T5/vienn/snps/'

files.base = list.files(path = path.base, pattern = ".fas")
base.pref = 'TAIR10_chr'

files.query = list.files(path = path.query, pattern = ".fasta")
query.name = gsub("*.fasta","",files.query)

query.remove = c('22001', '6191', '9764', '6237')

query.name.remain = c()
for(i.query in 1:length(query.name)){  # which accession to use
    if(query.name[i.query] %in% query.remove) next
  query.name.remain = c(query.name.remain, query.name[i.query])
}

```


# Read data
```{r}

# query.chr = 1
# base.chr = 1
# 
# i.query = 1  # which accession to use
# 
# base.file = paste0(base.pref, query.chr ,'.fas', collapse = '')
# query.file = files.query[i.query]
# pref.blsat.res = paste0(query.name[i.query], '_', query.chr, '_', collapse = '')
# 
# base.fas.fw = read.fasta(paste(path.base, base.file, sep = ''))
# query.fas = read.fasta(paste(path.query, query.file, sep = ''))
# 
# base.fas.fw = base.fas.fw[[base.chr]]
# base.fas.bw = reverseComplement(base.fas.fw)
# base.len = length(base.fas.fw)

# query.fas.chr = query.fas[[query.chr]]

# blast.res.file <- paste0(pref.blsat.res, as.character(base.chr), '.txt', collapse = '');

```


# read alignments and generate figures
```{r}


for(query.chr in 1:5){
  # base.chr = 1
  base.file = paste0(base.pref, query.chr ,'.fas', collapse = '')
  base.fas.fw = read.fasta(paste(path.base, base.file, sep = ''))
  base.fas.fw = base.fas.fw[[1]] # do not remove this 1, file contains only one sequence
  base.fas.bw = reverseComplement(base.fas.fw)
  base.len = length(base.fas.fw)
  plot.list = list()
  for(i.query in 1:length(query.name)){  # which accession to use
    
    pref.blsat.res = paste0(query.name[i.query], '_', query.chr, '_', collapse = '')
    blast.res.file <- paste0(pref.blsat.res, as.character(query.chr), '.rds', collapse = '');
    file.t = paste(path.t, blast.res.file, sep = '')
    
    
    t = readRDS(file.t)

    df = c()
    for(i in 1:nrow(t)) {
      if(t$dir[i] == 0){
        df = rbind(df, c(t[i, 'V2'], t[i, 'V4'], i, 0))
        df = rbind(df, c(t[i, 'V3'], t[i, 'V5'], i, 0))
      } else {
        df = rbind(df, c(t[i, 'V2'], base.len - t[i, 'V4'] + 1, i, 1))
        df = rbind(df, c(t[i, 'V3'], base.len - t[i, 'V5'] + 1, i, 1))
      }
        
    }
    
    df = as.data.frame(df)
    df$clr <- df$V3 %% 2;
    p <- ggplot(df, aes(x = V1, y=V2, color = as.factor(V4), group=as.factor(V3)  )) + 
      geom_line(show.legend = FALSE) + theme_bw() + theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) + ggtitle(query.name[i.query])
    # p
    plot.list[[i.query]] <- p
    pdf(paste(path.figures,  
              paste0('chr', as.character(query.chr), '_', pref.blsat.res, 
                     as.character(query.chr), '.pdf', collapse = ''), sep = ''), 
        width = 3.5, height = 3) 
    print(p)
    dev.off()
  }
  pdf(paste(path.figures,  paste0(as.character(query.chr), '.pdf', collapse = ''), sep = ''),
    width = 20, height = 21)
  ggarrange(plots=plot.list, nrow=7)
  dev.off()
}




```


# Coverage
```{r, fig.height=2, fig.width=2}

base.fac.all = list()
for(query.chr in 1:5){
    base.file = paste0(base.pref, query.chr ,'.fas', collapse = '')
    base.fas.fw = read.fasta(paste(path.base, base.file, sep = ''))
    base.fas.fw = base.fas.fw[[1]] # do not remove this 1, file contains only one sequence
    base.fac.all[[query.chr]] <- base.fas.fw
}
    
chr.coverage = c()
for(i.query in 1:length(query.name)){  # which accession to use
  if(query.name[i.query] %in% query.remove) next
  print(i.query)
  
  query.file = files.query[i.query]
  query.fas = read.fasta(paste(path.query, query.file, sep = ''))
  
  for(query.chr in 1:5){
    base.fas.fw = base.fac.all[[query.chr]]
    # base.fas.bw = reverseComplement(base.fas.fw)
    base.len = length(base.fas.fw)
    
    query.fas.chr = query.fas[[query.chr]]
  
  
    pref.blsat.res = paste0(query.name[i.query], '_', query.chr, '_', collapse = '')
    blast.res.file <- paste0(pref.blsat.res, as.character(query.chr), '.rds', collapse = '');
    file.t = paste(path.t, blast.res.file, sep = '')
    
    t = readRDS(file.t)
    
    t.base = t

    for(irow in 1:nrow(t.base)) {
      if(t.base[irow, 'dir'] == 1) {
        
        b1 <- t.base[irow,'V4']
        b2 <- t.base[irow,'V5']
        
        t.base[irow,'V4'] <- base.len - b1 + 1
        t.base[irow,'V5'] <- base.len - b2 + 1
      } 
    }
    
    # Coverage of base
    idx = rep(F, base.len)
    for(irow in 1:nrow(t.base)){
      idx[t.base[irow,'V4']:t.base[irow,'V5']] = T
    }
    cv.base = sum(idx) / base.len
    
    # Coverage of query
    idx = rep(F, length(query.fas.chr))
    for(irow in 1:nrow(t.base)){
      idx[t.base[irow,'V2']:t.base[irow,'V3']] = T
    }
    cv.query = sum(idx) / length(idx)
    
    chr.coverage = rbind(chr.coverage, c(cv.base, cv.query, query.chr, i.query))
  }
  # print(chr.coverage)
}

chr.coverage = as.data.frame(chr.coverage)
colnames(chr.coverage) <- c('tair10', 'accessions', 'chr')

p <- ggplot(chr.coverage, aes(x = tair10, color = as.factor(chr), fill = as.factor(chr) ))  + 
  geom_density(alpha=0.3)  + scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) + theme_minimal()
p 

pdf(paste(path.figures, 'coverage_tair10.pdf', sep = ''), 
    width = 4, height = 3) 
print(p)
dev.off()


p <- ggplot(chr.coverage, aes(x = accessions, color = as.factor(chr), fill = as.factor(chr) ))  + 
  geom_density(alpha=0.3)  + scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) + theme_minimal()
p 

pdf(paste(path.figures, 'coverage_accessions.pdf', sep = ''), 
    width = 4, height = 3) 
print(p)
dev.off()

```


# Get SNPs - then haploblocks
```{r}

# write('', file=paste0(path.snps, 'pacbio_snps.txt', collapse = ''), append=F)

for(query.chr in 2:5){
  base.file = paste0(base.pref, query.chr ,'.fas', collapse = '')
  base.fas.fw = read.fasta(paste(path.base, base.file, sep = ''))
  base.fas.fw = base.fas.fw[[1]] # do not remove this 1, file contains only one sequence
  base.fas.bw = reverseComplement(base.fas.fw)
  base.len = length(base.fas.fw)
  
  base.fas.fw = toupper(base.fas.fw)
  
  # ---------------------------------------------------------------------
  # Get positions for the alignment
  pos.all = c()
  for(i.query in 1:length(query.name)){  # which accession to use
    if(query.name[i.query] %in% query.remove) next
    print(i.query)
    pref.blsat.res = paste0(query.name[i.query], '_', query.chr, '_', collapse = '')
    blast.res.file <- paste0(pref.blsat.res, as.character(query.chr), '.rds', collapse = '');
    file.t = paste(path.t, blast.res.file, sep = '')
    
    t = readRDS(file.t)
   
    t.base = t
    
    # get alignments
    for(irow in 1:nrow(t.base)) {
      if(t.base[irow, 'dir'] == 1) {
        
        b1 <- t.base[irow,'V4']
        b2 <- t.base[irow,'V5']
        
        t.base[irow,'V4'] <- base.len - b2 + 1
        t.base[irow,'V5'] <- base.len - b1 + 1
        
        q1 <- t.base[irow,'V2']
        q2 <- t.base[irow,'V3']
    
        t.base[irow,'V2'] <- -q2
        t.base[irow,'V3'] <- -q1
        
        t.base[irow, 'V8'] <- reverseComplement(t.base[irow, 'V8'])
        t.base[irow, 'V9'] <- reverseComplement(t.base[irow, 'V9'])
      } 
    }
    
    aln.pos = rep(0, base.len)
    aln.nt = rep('-', base.len)
    
    for(irow in 1:nrow(t)) {
      s.q = strsplit(t[irow, 'V8'], '')[[1]]
      s.b = strsplit(t[irow, 'V9'], '')[[1]]
      
      # print(sum(s.q != s.b))
      
      pos.q = rep(0, t[irow, 'V7'])
      pos.q[s.q != '-'] <- t[irow, 'V2']:t[irow, 'V3']
      pos.b = t[irow, 'V4']:t[irow, 'V5']
      
      idx.b = which(s.b == '-')
      if(length(idx.b) > 0) {
        pos.q = pos.q[-idx.b]
        s.q = s.q[-idx.b]  
      }
        aln.pos[pos.b] = pos.q
        aln.nt[pos.b] = s.q
    } 
    
    snp.pos = which((aln.nt != '-') & (aln.nt != toupper(base.fas.fw)))
    flanking.match = 3
    idx = (snp.pos[-1] - snp.pos[-length(snp.pos)]) > flanking.match
    idx = c(idx, T) & c(T, idx)
    snp.pos = snp.pos[idx]

    # aln.all = rbind(aln.all, aln.nt)
    pos.all = c(pos.all, snp.pos)
    
  }
  pos.all = sort(unique(pos.all))
  
  # Get alignmnetn positions
  snp.pos = pos.all
  flanking.match = 3
  idx = (snp.pos[-1] - snp.pos[-length(snp.pos)]) > flanking.match
  idx = c(idx, T) & c(T, idx)
  pos.all = snp.pos[idx]
  
  
  # --------------------------------------------
  # Get alignment
  aln.all = c()
  for(i.query in 1:length(query.name)){  # which accession to use
    if(query.name[i.query] %in% query.remove) next
    print(i.query)
    pref.blsat.res = paste0(query.name[i.query], '_', query.chr, '_', collapse = '')
    blast.res.file <- paste0(pref.blsat.res, as.character(query.chr), '.rds', collapse = '');
    file.t = paste(path.t, blast.res.file, sep = '')
    
    t = readRDS(file.t)
   
    t.base = t
    
    # get alignments
    for(irow in 1:nrow(t.base)) {
      if(t.base[irow, 'dir'] == 1) {
        
        b1 <- t.base[irow,'V4']
        b2 <- t.base[irow,'V5']
        
        t.base[irow,'V4'] <- base.len - b2 + 1
        t.base[irow,'V5'] <- base.len - b1 + 1
        
        q1 <- t.base[irow,'V2']
        q2 <- t.base[irow,'V3']
    
        t.base[irow,'V2'] <- -q2
        t.base[irow,'V3'] <- -q1
        
        t.base[irow, 'V8'] <- reverseComplement(t.base[irow, 'V8'])
        t.base[irow, 'V9'] <- reverseComplement(t.base[irow, 'V9'])
      } 
    }
    
    aln.pos = rep(0, base.len)
    aln.nt = rep('-', base.len)
    
    for(irow in 1:nrow(t)) {
      s.q = strsplit(t[irow, 'V8'], '')[[1]]
      s.b = strsplit(t[irow, 'V9'], '')[[1]]
      
      # print(sum(s.q != s.b))
      
      pos.q = rep(0, t[irow, 'V7'])
      pos.q[s.q != '-'] <- t[irow, 'V2']:t[irow, 'V3']
      pos.b = t[irow, 'V4']:t[irow, 'V5']
      
      idx.b = which(s.b == '-')
      if(length(idx.b) > 0) {
        pos.q = pos.q[-idx.b]
        s.q = s.q[-idx.b]  
      }
        aln.pos[pos.b] = pos.q
        aln.nt[pos.b] = s.q
    } 
    aln.all = rbind(aln.all, aln.nt[pos.all])
  }

  
  tmp = cbind(rep(query.chr, length(pos.all)), pos.all, t(aln.all))
  colnames(tmp) = c('chr', 'pos', query.name.remain)
  write.table(x = tmp, file = paste0(path.snps, as.character(query.chr), '.txt', collapse = ''), quote = F,
              row.names = F)
  
  tmp.cut <- tmp[,colSums(tmp == '-') < 150000]
  tmp.cut <- tmp.cut[rowSums(tmp.cut == '-') < 5,]
  tmp.cut <- tmp.cut[rowSums(tmp.cut == '-') == 0,]
  write.table(x = tmp.cut, file = paste0(path.snps, as.character(query.chr), '_cut', '.txt', collapse = ''),
              quote = F,row.names = F)
  
  for(i in 3:ncol(tmp.cut)){
    tmp.cut[,i] = (tmp.cut[,i] != base.fas.fw[as.numeric(tmp.cut[,2])]) * 1
  }
  tmp.file = paste0(path.snps, as.character(query.chr), '_cut_num', 
                                         '.txt', collapse = '')
  write.table(x = tmp.cut, file = tmp.file, quote = F,row.names = F)
  
  tmp.cut = read.table(tmp.file, header = T, check.names = FALSE)
  maf = 0.1
  tmp.cut = tmp.cut[rowSums(tmp.cut[,3:ncol(tmp.cut)]) > maf * ncol(tmp.cut),]
  tmp.file.maf = paste0(path.snps, as.character(query.chr), '_cut_num_maf', 
                                         '.txt', collapse = '')
  write.table(x = tmp.cut, file = tmp.file.maf, quote = F,row.names = F)
  
  # transitions-transversions
  # tmp.nt.remains = tmp[tmp[,'pos'] %in% tmp.cut[,'pos'] ,]
  
  df = as.data.frame(tmp.cut[,'pos'])
  colnames(df) <- 'pos'
  df$pos = as.numeric(df$pos)
  p <- ggplot(df, aes(x = pos))  + 
    geom_density(alpha=0.3, adjust = 1/3) + theme_minimal() + 
    xlab(paste0('positions chr', query.chr, collapse = '')) + ylab('SNP density')
  p 
  
  pdf(paste(path.figures, paste0('snp_dens_chr', query.chr,'.pdf', collapse = ''), sep = ''), 
      width = 7, height = 3) 
  print(p)
  dev.off()
  
}








```


# Get Mobilome - versus TE - see remaing
```{r}




```






