---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

# Setup
```{r}
library(ggplot2)
library(Biostrings)
library('seqinr')
library('spgs')  # reverseComplement("actg")
source("/Users/anna/OneDrive/pushkin/cryptic/nanopore/scripts/synteny_infer.R")

path.base = '/Volumes/Samsung_T5/vienn/tair/'
path.query = '//Volumes/Samsung_T5/vienn/pacbio/'
path.blast.res <- '/Volumes/Samsung_T5/vienn/blast_res/'
```


# Read data
```{r}
base.file = 'TAIR10_chr1.fas'
query.file = 'A1.polished_contigs_scaffolded_softmasked_TAIR10.v1.1.fasta'

base.fas.fw = read.fasta(paste(path.base, base.file, sep = ''))
query.fas = read.fasta(paste(path.query, query.file, sep = ''))

base.fas.fw = base.fas.fw[[1]]

# base.chr.len = sapply(base.fas.fw, function(x) length(x))
# query.chr.len = sapply(query.fas, function(x) length(x))

# base.fas.fw = base.fas.fw[base.chr.len == max(base.chr.len)][[1]]
# query.fas = query.fas[query.chr.len == max(query.chr.len)][[1]]

# base.chr.name = names(base.chr.len)[base.chr.len == max(base.chr.len)]

base.fas.bw = reverseComplement(base.fas.fw)

```


# Read blast results
```{r}

query.chr = 1
blast.res.file <- paste0('A1_1_parts_', as.character(query.chr), '.txt', collapse = '');

t = read.table(paste(path.blast.res, blast.res.file, sep = ''), stringsAsFactors = F, header = F)

query.fas.chr = query.fas[[query.chr]]

# Remove duplicates
# t = t[!duplicated(t[,'V1']),]

# Remove short
t = t[t[,'V7'] > 500,]

# Remove low quality
quality.cutoff = 0.95
t = t[t[,'V6'] > quality.cutoff,]

# direction
t$dir = c()
for(irow in 1:nrow(t)) {
  if(t[irow,'V5'] < t[irow,'V4']) {
    
    pos1 = t[irow,'V5']
    pos2 = t[irow,'V4']
    
    t[irow,'V5'] <- length(base.fas.bw) - pos1 + 1
    t[irow,'V4'] <- length(base.fas.bw) - pos2 + 1
    t[irow, 'dir'] = 1
  } else {
    t[irow, 'dir'] = 0
  }
}

# Get right positions
start.pos = as.numeric(sapply(strsplit(t[,1], "\\|"), "[", 4)) - 1
t[,2:3] = t[,2:3] + start.pos

rownames(t) <- NULL

# check
for(irow in 1:nrow(t)) {
  if(nchar(t[irow,'V8']) != nchar(t[irow,'V9'])) stop('aaa')
}

source("/Users/anna/OneDrive/pushkin/cryptic/nanopore/scripts/synteny_infer.R")
checkCorrespToGenome(t, query.fas = query.fas.chr, 
                     base.fas.fw = base.fas.fw, 
                     base.fas.bw = base.fas.bw)
showt(t)


```




# Start
```{r}
# t = read.table('/Volumes/Samsung_T5/vienn/tmp/tmp4.txt', stringsAsFactors = F)
# plot(t[,'V4'], t[,'V2'])
# # orderliness
# 
# idx = abs(t[,'V2'] - t[,'V4']) < 10^6
# t <- t[idx,]
# rownames(t) <- NULL
# 
# rownames(t) <- NULL
# 
# # remove outliers - single
# for(i in 1:2) {
#   a1 = t[1:(nrow(t)-2),'V4']
#   a2 = t[2:(nrow(t)-1),'V4']
#   a3 = t[3:(nrow(t)),'V4']
#   
#   bidx = c(F, abs(a3-a1) < abs((a2*2 - a1 - a3))/2, F);
#   idx = 1:nrow(t);
#   idx[bidx]
#   t <- t[-idx[bidx],];
#   rownames(t) <- NULL
# }
# 
# plot(t[,'V4'], t[,'V2'])
# 
# 
```


# First glue
```{r}

t = t[order(t[,'V4']),]

ipos = 1
while(ipos < nrow(t)) {
  # print(ipos)
  idx = which((t[,'V4'] == (t[ipos,'V5'] + 1)) & (t[,'V2'] == (t[ipos,'V3'] + 1)))
  if (length(idx) == 0){
    ipos = ipos+1
    next
  }
  
  # if(idx != (ipos+1)) print(idx)  # glue with not the next record
  
  t[ipos, 'V3'] <- t[idx, 'V3']
  t[ipos, 'V5'] <- t[idx, 'V5']
  t[ipos, 'V7'] <- t[ipos, 'V7'] + t[idx, 'V7']
  t[ipos, 'V8'] <- paste(t[ipos, 'V8'], t[idx, 'V8'], sep = '')
  t[ipos, 'V9'] <- paste(t[ipos, 'V9'], t[idx, 'V9'], sep = '')
  t <- t[-idx,]
}


rownames(t) <- NULL
t0 = t

```


## Distribution of gaps
```{r}

plot(t[,'V4'], t[,'V2'])

hist(log(t[2:(nrow(t)),'V2'] - t[1:(nrow(t)-1),'V3'], base = 10), 100)

sum(t[2:(nrow(t)),'V2'] - t[1:(nrow(t)-1),'V3'] < 10^3)
intersect(which(t[2:(nrow(t)),'V2'] - t[1:(nrow(t)-1),'V3'] > 10^4),
          which(t[2:(nrow(t)),'V4'] - t[1:(nrow(t)-1),'V5'] > 10^4))

setdiff(which(t[2:(nrow(t)),'V2'] - t[1:(nrow(t)-1),'V3'] > 10^4),
          which(t[2:(nrow(t)),'V4'] - t[1:(nrow(t)-1),'V5'] > 10^4))

```

# pairwise alignment
```{r}

file.log <- '/Users/anna/OneDrive/vienn/pacbio/log.txt'
write('', file=file.log, append=F)

thresholds = c(100, 1000, 2000)
# thresholds = c(1000)

for(threshold in thresholds) {

  irow = 1
  while(irow < nrow(t)) {
    # print(threshold)
    # print(irow)
    
    # if a base fragment of irow is already within some base fragment
    idx = (t[,'V4'] <= t[irow,'V4']) & (t[,'V5'] >= t[irow,'V5'])
    idx[irow] = F
    if(sum(idx) > 0){
      irow <- irow + 1
      next
    }
    
    # # if a query fragment of irow is already within some query fragment
    # idx = (t[,'V2'] <= t[irow,'V2']) & (t[,'V3'] >= t[irow,'V3'])
    # idx[irow] = F
    # if(sum(idx) > 0){
    #   t <- t[-irow,]
    #   next
    # }
    
    # idx = abs(t[,'V4'] - t[irow,'V5']) + abs(t[,'V2'] - (t[irow,'V3']))
    idx = abs(t[,'V2'] - (t[irow,'V3']))
    idx[irow] = 10^10
    idx[t[,'V2'] <= t[irow,'V2']] = 10^10
    idx[t[,'V4'] <= t[irow,'V4']] = 10^10
    
    # If new fragment is in irow fragment by query
    idx.inside = (t[,'V2'] >= t[irow,'V2']) & (t[,'V3'] <= t[irow,'V3'])
    idx[idx.inside] = 10^10
    
    idx[t$dir != t$dir[irow]] = 10^10
    
    idx = which(idx == min(idx))
    idx = idx[1]
    
    # if(t[idx,'V1'] == 'part|35-1|805001|806000') spot('find')
    
    flag = (abs(t[idx, 'V2'] - t[irow, 'V3']) < threshold) & 
      (abs(t[idx, 'V4'] - t[irow, 'V5']) < threshold)
    # if(irow == 29) stop('aa')
    
    
    if(!flag) {
      irow <- irow + 1
      next
    }
    
    if(threshold == 10^4){
      n.char = 50
      n.char.thresh = 100  
    } else {
      n.char = 9
      n.char.thresh = 50
    }
    
    
    d.tmp = 0
    while (d.tmp < n.char.thresh) {
      n.char = n.char + 1
      pos.query.start = max(1, t[irow, 'V3'] - n.char + sumGapLastN(t[irow, 'V8'], n.char))
      pos.query.end = t[idx, 'V2'] + n.char - sumGapFirstN(t[idx, 'V8'], n.char)
      
      
      pos.base.start = max(1, t[irow, 'V5'] - n.char + sumGapLastN(t[irow, 'V9'], n.char))
      pos.base.end = t[idx, 'V4'] + n.char - sumGapFirstN(t[idx, 'V9'], n.char)
      
      d.tmp = min(pos.query.end - pos.query.start, pos.base.end - pos.base.start)
    }
    
    if(t$dir[irow] == 0){
      base.fas = base.fas.fw
    } else {
      base.fas = base.fas.bw
    }
    
    seq.query <- toupper(query.fas.chr[(pos.query.start+1):(pos.query.end-1)])
    seq.base <- toupper(base.fas[(pos.base.start+1):(pos.base.end-1)])  
    
    globalAlign<- pairwiseAlignment(paste0(seq.query, collapse=''), 
                                    paste0(seq.base, collapse=''), gapOpening = 20, type="global")
    globalAlign 
    
    
    write(paste((pos.query.start+1), (pos.query.end-1), as.character(alignedPattern(globalAlign))),
          file=file.log, append=TRUE)
    write(paste((pos.base.start+1), (pos.base.end-1), as.character(alignedSubject(globalAlign))), 
          file=file.log, append=TRUE)
    write('\n', file=file.log, append=TRUE)
    
    
    seq.query.new = paste(removeLastN(t[irow, 'V8'], n.char),  
                          alignedPattern(globalAlign), 
                          removeFirstN(t[idx, 'V8'], n.char), sep = '')
    
    seq.base.new = paste(removeLastN(t[irow, 'V9'], n.char),  
                         alignedSubject(globalAlign), 
                         removeFirstN(t[idx, 'V9'], n.char), sep = '')
    
    if(nchar(seq.query.new) != nchar(seq.base.new)) stop('aa')
    
    t1 <- t
    
    t[irow, 'V8'] <- seq.query.new
    t[irow, 'V9'] <- seq.base.new
    
    t[irow, 'V3'] <- t[idx, 'V3']
    t[irow, 'V5'] <- t[idx, 'V5']
    t[irow, 'V7'] <- nchar(seq.base.new)
    
    t <- t[-idx,]
    
    # s1 = seq.base.new
    # s2 = paste0(base.fas.fw[t[irow,'V4']:t[irow,'V5']], collapse = '')
    
    s.base = strsplit(seq.base.new,'')[[1]]
    idx.nogap.base = which(s.base != '-')
    if(length(idx.nogap.base) != abs(t[irow,'V5'] - t[irow,'V4']) + 1) stop('wrong in base')
    
    s.query = strsplit(seq.query.new,'')[[1]]
    idx.nogap.query = which(s.query != '-')
    if(length(idx.nogap.query) != abs(t[irow,'V3'] - t[irow,'V2']) + 1) stop('wrong in query')
    
    
    if(idx < irow) {
      irow = irow - 1
    }    
    
    rownames(t) <- NULL
  }
}

showt(t, c(irow, idx))

t1 = t
```


## Remove outliers
```{r}

plot(t[, 'V4'],t[, 'V2'])

fragment.len.cutoff = 2000
plot(t[t[,'V7'] <= fragment.len.cutoff, 'V4'], 
     t[t[,'V7'] <= fragment.len.cutoff, 'V2'])

t = t[t[,'V7'] > fragment.len.cutoff,]
plot(t[, 'V4'],t[, 'V2'])

hist(log(t[,'V7'], base = 10), 100)

rownames(t) <- NULL


```


# if something is inside something in query - remove it
```{r}
irow = 1

while(irow < nrow(t)) {

  # index of all, that overlap with the current one (irow) in the right side
  idx = setdiff(which((t[,'V2'] >= t[irow,'V2']) & (t[,'V3'] <= t[irow,'V3'])), irow)
  if(length(idx) > 0) {
    stop('a')
   t = t[-idx,] 
   irow = irow - sum(idx < irow)
  }
    
  # # index of all, that overlap with the current one (irow) in the right side
  # idx = setdiff(which((t[,'V2'] >= t[irow,'V2']) & (t[,'V2'] <= t[irow,'V3'])), irow)
  # if(length(idx) > 0) {
  #   stop('a')
  #  t = t[-idx,] 
  #  irow = irow - sum(idx < irow)
  # }
  
  # # index of all, that overlap with the current one (irow) in the left side
  # idx = setdiff(which((t[,'V3'] >= t[irow,'V2']) & (t[,'V3'] <= t[irow,'V3'])), irow)
  # if(length(idx) > 0) {
  #   stop('b')
  #   t = t[-idx,] 
  #   irow = irow - sum(idx < irow)
  # }
  irow = irow + 1
}

showt(t, c(irow, idx))

```

# Plot syntheny blocks
```{r fig.width=3, fig.height=3}
df = c()
for(i in 1:nrow(t)) {
  if(t$dir[i] == 0){
    df = rbind(df, c(t[i, 'V2'], t[i, 'V4'], i))
    df = rbind(df, c(t[i, 'V3'], t[i, 'V5'], i))
  } else {
    df = rbind(df, c(t[i, 'V2'], length(base.fas.fw) - t[i, 'V4'] + 1, i))
    df = rbind(df, c(t[i, 'V3'], length(base.fas.fw) - t[i, 'V5'] + 1, i))
  }
    
}

df = as.data.frame(df)
df$V4 <- df$V3 %% 2;
p <- ggplot(df, aes(x = V1, y=V2, color = as.factor(V4), group=as.factor(V3)  )) + 
  geom_line(show.legend = FALSE) + theme_bw()
p

```


# Get something to glue
```{r}
t <- t[order(t[,'V']),]


```




# Mobilome
```{r}
mobilome <- c()

for(irow in 1:nrow(t)) {
# for(irow in 1:50) {
  s.query <- strsplit(t[irow, 'V8'], '')[[1]]
  s.base <- strsplit(t[irow, 'V9'], '')[[1]]
  s <- c()
  f <- ''
  for(icol in 1:length(s.query)){
    if(s.query[icol] == '-') {
      if(f == 'query'){
        if(length(s) > 0) mobilome <- c(mobilome, paste0(s, collapse =''))
        s <- c()
      }
      f = 'base' # get sequence from base
      s <- c(s, s.base[icol])
    } else if(s.base[icol] == '-') {
      if(f == 'base'){ 
        if(length(s) > 0) mobilome <- c(mobilome, paste0(s, collapse =''))
        s <- c()
      }
      f = 'query'
      s <- c(s, s.query[icol])
    } else if (length(s) > 0) {
      # stop('aa')
      # if (length(s) == 7) {
      #   stop('aa')
      # }
      mobilome <- c(mobilome, paste0(s, collapse =''))
      s <- c()
      f = ''
      

      
    }
    

  }

}


```


```{r}
df = c()
for(i in 1:nrow(t)) {
  df = rbind(df, c(t[i, 'V2'], t[i, 'V3'], i))
  df = rbind(df, c(t[i, 'V4'], t[i, 'V5'], i))
}


df = as.data.frame(df)
library(ggplot2)
ggplot(df, aes(x = V1, y=V2, color = as.factor(V3)  )) + geom_line(show.legend = FALSE) 



```




```{r}

mobilome.t <- table(mobilome)
mobilome.unique.len <- sapply(names(mobilome.t), nchar)
mobilome.t = mobilome.t[order(mobilome.unique.len)]
mobilome.unique.len <- sapply(names(mobilome.t), nchar)


mobilome.t[]

plot(log(mobilome.t, base = 2))



idx <- (mobilome.t > 2^2) & (mobilome.t < 2^6) & 
  (rank(mobilome.unique.len) > length(mobilome.t)/3)

mobilome.t[idx]


mobilome.t[(mobilome.unique.len>15) & (mobilome.unique.len<100)]


for(len in 15:50){
  mobilome.prev = names(mobilome.t[mobilome.unique.len == (len-1)])
  
  tmp = mobilome.t[mobilome.unique.len == (len-1)]
  tmp[tmp >1]
  
  mobilome.prev = names(tmp[tmp >1])
  
  scores = c()
  for(i in 1:length(mobilome.prev)) {
    for(j in i:length(mobilome.prev)) {
      if (i == j) next
        seq1 <- mobilome.prev[i]
        seq2 <- mobilome.prev[j]
        globalAlign<- pairwiseAlignment(seq1, seq2)
        scores = rbind(scores, c(i, j, globalAlign@score))
    }
  }
}

```



```{r}
nts <- c('A', 'C', 'G', 'T')
cnt <- mobilome.t[nts]
freq <- cnt / sum(cnt)

cmb <- gtools::permutations(4, 7, repeats.allowed=TRUE)
nts.cmb <- sapply(1:nrow(cmb), function(i) paste0(nts[cmb[i,]], collapse = ''))

cnt.cmb <- mobilome.t[nts.cmb]

freq.cmb <- sapply(1:nrow(cmb), function(i) prod(freq[cmb[i,]]))

plot(freq.cmb, cnt.cmb)

text(freq.cmb, cnt.cmb, nts.cmb, cex=0.5, pos=3,col="red", adj = c(0,1))

```


```{r}
d<-head(mtcars)
plot(d[,'wt'], d[,'mpg'], 
     main="Milage vs. Car Weight\n~~~~~~~~~~~~~~~~~~~",
      xlab="Weight", ylab="Miles/(US) gallon",
      pch=19, col="darkgreen")
text(d[,'wt'], d[,'mpg'],  row.names(d),
     cex=0.65, pos=3,col="red") 
```


<!-- ```{r} -->
<!-- # find the largest list, which is sorted. -->

<!-- hist(abs(t[2:nrow(t),'V4'] - t[1:(nrow(t) - 1),'V4'])) -->

<!-- plot(t[,'V4'], t[,'V2']) -->

<!-- t = t[order(t[,'V4']),] -->


<!-- base.pos = list() -->
<!-- for(imatch in 1:nrow(t)) { -->

<!--   query.id <- t[imatch, 'V2']:t[imatch, 'V3'] -->
<!--   base.id <- t[imatch, 'V4']:t[imatch, 'V5'] -->

<!--   t[imatch, 'V8'] = toupper(t[imatch, 'V8']) -->
<!--   t[imatch, 'V9'] = toupper(t[imatch, 'V9']) -->

<!--   query.seq = strsplit(t[imatch, 'V8'], "")[[1]] -->
<!--   base.seq = strsplit(t[imatch, 'V9'], "")[[1]] -->

<!--   base.id.nongap <- which(!(base.seq %in% c("-", 'N'))) -->

<!--   # query.id.gap <- which(query.seq %in% c("-", 'N')) -->
<!--   # base.id.gap <- which(base.seq %in% c("-", 'N')) -->
<!--   # if(length(query.id.gap) + length(query.id) != length(base.id.gap) + length(base.id)) stop('problem!!!') -->

<!--   query.id.aln = rep(0, t[imatch, 'V7']) -->
<!--   query.id.aln[which(!(query.seq %in% c("-", 'N')))] <- query.id -->

<!--   query.id.in.base = query.id.aln[base.id.nongap] -->
<!--   base.seq.in.base = base.seq[base.id.nongap] -->

<!--   query.seq.in.base = query.seq -->
<!--   query.seq.in.base[query.seq == base.seq] = '' -->
<!--   query.seq.in.base = query.seq.in.base[base.id.nongap] -->


<!--   base.nt = paste(base.id, base.seq.in.base, sep = ':') -->

<!--   query.nt = paste(query.id.in.base, query.seq.in.base, sep = ':') -->
<!--   # query.nt[query.id.in.base != (c(0, query.id.in.base[1:(length(query.id.in.base)-1)]) + 1)] -->

<!--   base.tmp <- list() -->
<!--   base.tmp[base.nt] <- query.nt -->

<!--   base.pos <- c(base.pos, base.tmp) -->
<!--   # base.pos <- base.pos[!duplicated(base.pos)] -->
<!--   print(length(base.pos)) -->
<!-- } -->



<!-- dup.names = names(base.pos)[duplicated(names(base.pos))] -->
<!-- dup.name = dup.names[1] -->

<!-- base.pos[names(base.pos) %in% dup.name] -->
<!-- ``` -->



<!-- ```{r} -->
<!-- te <- read.table('/Volumes/Samsung_T5/vienn/base/te/TAIR10_Transposable_Elements.txt', header = 1) -->
<!-- te <- te[order(te$Transposon_min_Start),] -->
<!-- ``` -->



